src/unix/linux-core.c

```diff
@@ -725,27 +732,383 @@ uint64_t uv_get_total_memory(void) {
+  if (CGROUPS_VERSION_UNKNOWN == info->cgroups_version) {
+    goto cleanup;
+  }
+  
+  /* Read /proc/self/cgroup to get hierarchy path. */
+
+  fp = uv__open_file(PROC_SELF_CGROUP);
+  if (fp == NULL) {
+    rc = UV__ERR(errno);
+    goto cleanup;
+  }
+  
+  hierarchy_path = uv__malloc(UV__PATH_MAX);
+  /* + 3 for two colons, and terminal character. */
+  subsystem_search_string = uv__malloc(strlen(subsystem) + 3);
+  if (hierarchy_path == NULL || subsystem_search_string == NULL) {
+    rc = UV_ENOMEM;
+    goto cleanup;
+  }
+
+  hierarchy_path[0] = '\0';
+  snprintf(subsystem_search_string, strlen(subsystem) + 3, ":%s:", subsystem);
+
+  while (feof(fp) == 0) {
+    if (getline(&buf, &buf_size, fp) < 0) {
+      if (feof(fp) != 0)
+        break;
+      rc = UV_EIO;
+      goto cleanup;
+    }
+    buf_strlen = strlen(buf);
+    if ('\n' == buf[buf_strlen - 1])
+      buf[buf_strlen - 1] = '\0';
+
+    if (CGROUPS_VERSION_1 == info->cgroups_version) {
+      hierarchy_path_search_ptr = strstr(buf, subsystem_search_string);
+      if (hierarchy_path_search_ptr != NULL)
+        hierarchy_path_search_ptr += strlen(subsystem_search_string);
+    } else { /* if (CGROUPS_VERSION_2 == info->cgroups_version) */
+      /* 3 is the string length of "0::". */
+      if (strncmp(buf, "0::", 3) == 0)
+        hierarchy_path_search_ptr = buf + 3;
+    }
+    if (hierarchy_path_search_ptr != NULL) {
+      if (uv__strscpy(hierarchy_path,
+                      hierarchy_path_search_ptr,
+                      UV__PATH_MAX - 1) < 0) {
+        rc = UV_E2BIG;
+        goto cleanup;
+      }
+      break;
+    }
+  }
+
+  fclose(fp);
+  fp = NULL;
+
+  /*
+   * The hierarchy path should be prefixed with the root path from mountinfo,
+   * and should be replaced with the mount point.
+   */
+  if (strncmp(hierarchy_path, root, strlen(root)) == 0) {
+    size_t path_size;
+    hierarchy_path_inner = hierarchy_path + strlen(root);
+    /* +2 for "/" and null terminator. */
+    path_size = strlen(mount_point) + strlen(hierarchy_path_inner) + 2;
+    info->path = uv__malloc(path_size);
+    if (info->path == NULL)
+      rc = UV_ENOMEM;
+    else
+      snprintf(info->path,
+               path_size,
+               "%s/%s",
+               mount_point,
+               hierarchy_path_inner);
+  } else {
+    info->cgroups_version = CGROUPS_VERSION_UNKNOWN;
+  }
+
+
+cleanup:
+  if (fp != NULL)
+    fclose(fp);
+  /* buf is (re-)allocated via getline, so use standard free. */
+  if (buf != NULL)
+    free(buf);
+  if (root != NULL)
+    uv__free(root);
+  if (mount_point != NULL)
+    uv__free(mount_point);
+  if (hierarchy_path != NULL)
+    uv__free(hierarchy_path);
+  if (subsystem_search_string != NULL)
+    uv__free(subsystem_search_string);
+  return rc;
+}
+
+
+static uint64_t uv__read_cgroups_uint64(const char* path, const char* param) {
+
+  char filename[UV__PATH_MAX];
   char buf[32];  /* Large enough to hold an encoded uint64_t. */
   uint64_t rc;
```

Comment:
`rc` is now undefined on some code paths
```suggestion
  uint64_t rc = 0;
```